algorithms/src/std_algorithms/impl/Kokkos_RandomAccessIterator.hpp

```diff
@@ -111,77 +118,87 @@ class RandomAccessIterator< ::Kokkos::View<DataType, Args...> > {
 
   KOKKOS_FUNCTION
   reference operator[](difference_type n) const {
-    return m_view(m_current_index + n);
+    if constexpr (is_contiguous)
+      return *(m_data + n);
+    else
+      return *(m_data + n * m_stride);
   }
 
   KOKKOS_FUNCTION
   iterator_type& operator+=(difference_type n) {
-    m_current_index += n;
+    if constexpr (is_contiguous)
+      m_data += n;
+    else
+      m_data += n * m_stride;
     return *this;
   }
 
   KOKKOS_FUNCTION
   iterator_type& operator-=(difference_type n) {
-    m_current_index -= n;
+    if constexpr (is_contiguous)
+      m_data -= n;
+    else
+      m_data -= n * m_stride;
     return *this;
   }
 
   KOKKOS_FUNCTION
   iterator_type operator+(difference_type n) const {
-    return iterator_type(m_view, m_current_index + n);
+    auto it = *this;
+    it += n;
+    return it;
   }
 
   KOKKOS_FUNCTION
   iterator_type operator-(difference_type n) const {
-    return iterator_type(m_view, m_current_index - n);
+    auto it = *this;
+    it -= n;
+    return it;
   }
 
   KOKKOS_FUNCTION
   difference_type operator-(iterator_type it) const {
-    return m_current_index - it.m_current_index;
+    if constexpr (is_contiguous)
+      return m_data - it.m_data;
+    else
+      return (m_data - it.m_data) / m_stride;
   }
 
   KOKKOS_FUNCTION
-  bool operator==(iterator_type other) const {
-    return m_current_index == other.m_current_index &&
-           m_view.data() == other.m_view.data();
-  }
+  bool operator==(iterator_type other) const { return m_data == other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator!=(iterator_type other) const {
-    return m_current_index != other.m_current_index ||
-           m_view.data() != other.m_view.data();
-  }
+  bool operator!=(iterator_type other) const { return m_data != other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator<(iterator_type other) const {
-    return m_current_index < other.m_current_index;
-  }
+  bool operator<(iterator_type other) const { return m_data < other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator<=(iterator_type other) const {
-    return m_current_index <= other.m_current_index;
-  }
+  bool operator<=(iterator_type other) const { return m_data <= other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator>(iterator_type other) const {
-    return m_current_index > other.m_current_index;
-  }
+  bool operator>(iterator_type other) const { return m_data > other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator>=(iterator_type other) const {
-    return m_current_index >= other.m_current_index;
-  }
+  bool operator>=(iterator_type other) const { return m_data >= other.m_data; }
+
+  KOKKOS_FUNCTION
+  reference operator*() const { return *m_data; }
 
   KOKKOS_FUNCTION
-  reference operator*() const { return m_view(m_current_index); }
+  pointer data() const { return m_data; }
```

Comment:
OK, so I will make some general comments.
1. You are changing the semantics of the random iterator. It was sharing the ownership of the view, and your changes are such that this is no longer the case. I guess no one was actually relying on this and I could even reason about it and say that it is quite stupid that this iterator asks for the ownership.
2. Therefore, I wonder if this is just a `mdspan` (it has a pointer, and a stride). Not sure how you'd iterate over it though but food for thought.

Reply:
1. Correct. I think it makes iterator behave more like the standard std iterators.
2. I think `mdspan` has more data than that.

Reply:
@romintomasetti: even reason about it and say that it is quite stupid that this iterator asks for the ownership.
There are nicer ways to disagree with something. 
you are basically calling stupid me and the other people who worked on this class before (which is not just me).

Reply:
@aprokop @crtrott @dalg24 @fnrizzi 

I would like to seriously reconsider the `mdspan` approach, at least for the specialization `is_always_stride = true`.

I think the `mdspan` implementation would be much clearer and maintainable, and I bet that with `-O3` flag, most compilers will be able to optimize away all the unnecessary things from it and make as efficient as the current `ptr + stride` "manual" option.

See also my comment below https://github.com/kokkos/kokkos/pull/7304#discussion_r1824040479.

Note: I know the `Kokkos` implementation of `mdspan` is enabled by default (`DKokkos_ENABLE_IMPL_MDSPAN`) but is disabled for a few configurations.